rir/src/compiler/opt/load_elision.cpp

```diff
@@ -47,8 +56,13 @@ struct AvailableLoads : public StaticAnalysis<IntersectionSet<ALoad>> {
         AbstractResult res;
         if (auto ld = LdVar::Cast(i)) {
             for (auto& ald : state.available) {
-                if (ald.same(ld))
-                    return AbstractResult::None;
+                if (ald.same(ld)) {
+                    if (ld->type.isA(ald.type)) {
+                        ald.type = ld->type;
```

Comment:
maybe I am stupid now. But to me this looks like you are just storing the type of whatever instruction the analysis saw last. given the analysis does not have a fixed order this could be fairly random. In the end you might pick up a more precise type or also a less precise one.

If I understand you correctly you wanted to prevent this pass from degrading to a less precise type. Which seems like a reasonable thing to do. So why don't you do just that on the replacement side instead of in the analysis?

Reply:
You’re right that I’m trying to store the most precise type. What this should do is, only store the type of the last-seen instruction if it’s a subtype of the current type.

Say the analysis finds `int %1 = Ldvar a` and then `val %2 = LdVar a` and both loads are the same. The `ALoad` will have type `int` when it sees `val %2`, and since `val->isA(int)` is false, the type will stay `int`.

Of course, if one type is `int` and the other is `real` (neither is a subtype), the final type will be that of the first instruction, since neither type is more precise. But I don’t really know how to fix this since we don’t have true intersection types, and this is already what would happen.

Reply:
i see, I missed the test. ok